### PR TITLE
Optimized numpy functions in Converter

### PR DESCRIPTION
**Pull request by kilroythethird**
This PR optimized the conversion speed.
This is done mainly by choosing more performant variants for the numpy processing in the `Converter` class.

The Converter now also gets whole batches from the `Predict.queue_out_frames` method. This was mainly done to enable batch processing and potential replacing numpy functions with keras ones later on, but it also increased my speed by about 3it/s.

Optimizing the worker Plugins remains to be done. They did not take enough time in my setup to be really relevant. This may be different with other settings.

Benchmarks:
- Current staging: 16.45 it/s
- This PR WITHOUT providing whole batches to the converter: 21.40 it/s
- This PR as is: 24.69 it/s

Tested on 10k frames with a resolution of 1280x720.
The tested model on my system can do raw prediction (without anything else) with about 37.2 it/s

### CODE REVIEW

File: lib/convert.py [hunk 1]
```diff
@@ -178,7 +183,7 @@ def get_image_mask(self, new_face, detected_face, predicted_mask):
         else:
             logger.trace("Adding mask to alpha channel")
             new_face = np.concatenate((new_face, mask), -1)
-        new_face = np.clip(new_face, 0.0, 1.0)
+        new_face = np.clip(new_face, 0.0, 1.0, out=new_face)
```
[the remark this answers is not in the document]
Comment (kilroythethird) [reply lib/convert.py]: Its already using inplace modifications.
But you are right on that we don't need to save return value. It won't do any harm tho.

File: lib/convert.py [hunk 2]
```diff
@@ -115,7 +115,12 @@ def patch_image(self, predicted):
         new_image = self.get_new_image(predicted, frame_size)
         patched_face = self.post_warp_adjustments(predicted, new_image)
         patched_face = self.scale_image(patched_face)
-        patched_face = np.rint(patched_face * 255.0).astype("uint8")
+        patched_face *= 255.0
+        patched_face = np.rint(
+            patched_face,
+            out=np.zeros(patched_face.shape, dtype="uint8"),
```
[the remark this answers is not in the document]
Comment (kilroythethird) [reply lib/convert.py]: Thx.
True, altho the performance difference between zeros and empty is minimal.
Changing to empty.

File: lib/convert.py [hunk 3]
```diff
@@ -126,10 +131,10 @@ def get_new_image(self, predicted, frame_size):
         logger.trace("Getting: (filename: '%s', faces: %s)",
                      predicted["filename"], len(predicted["swapped_faces"]))
 
-        placeholder = predicted["image"] / 255.0
-        placeholder = np.concatenate((placeholder,
-                                      np.zeros((frame_size[1], frame_size[0], 1))),
-                                     axis=-1).astype("float32")
+        placeholder = np.zeros((frame_size[1], frame_size[0], 4), dtype="float32")
+        placeholder[:, :, :3] = predicted["image"]
```
[the remark this answers is not in the document]
Comment (kilroythethird) [reply lib/convert.py]: Doing the division by 255 as an inplace operation (x /= 255.) is faster than creating a temporary new array by doing x = x / 255.

I think we need the alpha channel in placeholder at that point as we use it to write the faces (including alpha channel) in cv2.warpAffine to it.
I agree tho that adding the alpha channel and then split it off later on is a bit silly, but i don't really see any other way.

[the remark this answers is not in the document]
Comment (kvrooman) [reply lib/convert.py]: this was the suggestion, but honestly any gains are small as i nentioned

~~placeholder[:, :, :3] = predicted["image"]~~
pred_image = predicted["image"].astype("float32")
pred_image /= 255.
placeholder[:, :, :3] = pred_image
....
~~background = predicted["image"][:, :, :3].astype("float32")~~
~~background /= 255.0~~
background = pred_image

[the remark this answers is not in the document]
Comment (kilroythethird) [reply lib/convert.py]: Oh, i see. Good catch. Getting rid of a cast+copy and a division is nothing to sneeze at in this context.
Will add this tomorrow

**Comment by kilroythethird** [comment]
Added kvroomans suggestions.
Should be good to go

**Pull request merged**